src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/StreamFormatter.cs

```diff
@@ -254,6 +255,78 @@ internal MessageBodyStream(Message message, string wrapperName, string wrapperNs
                 _isRequest = isRequest;
             }
 
+            public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                EnsureStreamIsOpen();
+                if (buffer == null)
+                    throw TraceUtility.ThrowHelperError(new ArgumentNullException("buffer"), _message);
+                if (offset < 0)
+                    throw TraceUtility.ThrowHelperError(new ArgumentOutOfRangeException("offset", offset,
+                                                    SR.Format(SR.ValueMustBeNonNegative)), _message);
+                if (count < 0)
+                    throw TraceUtility.ThrowHelperError(new ArgumentOutOfRangeException("count", count,
+                                                    SR.Format(SR.ValueMustBeNonNegative)), _message);
+                if (buffer.Length - offset < count)
+                    throw TraceUtility.ThrowHelperError(new ArgumentException(SR.Format(SR.SFxInvalidStreamOffsetLength, offset + count)), _message);
+
+                try
+                {
+                    if (_reader == null)
+                    {
+                        _readAheadStream =
+                            _message.Properties[ReadAheadWrappingStream.ReadAheadWrappingStreamPropertyName] as ReadAheadWrappingStream;
+                        // Fill buffer so reading to the body contents shouldn't cause a read through to transport stream
+                        if (_readAheadStream != null)
```

Comment:
Under what circumstances could this be null?  Should we assert or throw if it is?  Or is the intent that we softly use it when it is present but tolerate when it is not?  What kind of scenario test could be written to verify correct behavior in the null case?